pystac/link.py

```diff
@@ -158,16 +158,42 @@ def get_absolute_href(self) -> Optional[str]:
             from this link; however, if the link is relative, has no owner,
             and has an unresolved target, this will return a relative HREF.
         """
-        if self.is_resolved():
-            href = cast(pystac.STACObject, self.target).get_self_href()
+        if self._target_object:
+            href = self._target_object.get_self_href()
         else:
-            href = cast(Optional[str], self.target)
+            href = self._target_href
 
         if href is not None and self.owner is not None:
             href = make_absolute_href(href, self.owner.get_self_href())
 
         return href
 
+    @property
+    def target(self) -> Union[str, "STACObject_Type"]:
+        """The target of the link. If the link is unresolved, or the link is to something
+        that is not a STACObject, the target is an HREF. If resolved, the target is a
+        STACObject."""
+        if self._target_object:
+            return self._target_object
+        elif self._target_href:
+            return self._target_href
+        else:
+            raise ValueError("No target defined for link.")
+
+    @target.setter
+    def target(self, target: Union[str, "STACObject_Type"]) -> None:
+        """Sets this link's target to a string or a STAC object."""
+        if isinstance(target, str):
+            self._target_href = target
+            self._target_object = None
+        else:
+            self._target_href = None
+            self._target_object = target
+
+    def get_target_str(self) -> Optional[str]:
+        """Returns this link's target as a string."""
```

Comment:
Perhaps this should be worded differently - if you create a Link with the target that is a STACObject, this would imply to me that it would return the str representation of the link, i.e. call `get_self_href` on the target; it doesn't and would return None.

Or perhaps this should, and if _target_href is None and there's a _target_object, to return _target_object.get_self_href?

Reply:
Implemented in 3561a5c718471da5716f52f917698089b84edd59. I _think_ someone could infinitely recurse themselves if they created a self link with only a target object (no string href) but that'd be a weird thing to do.

Reply:
Added recursion check in 21feb889d0e05fd29a96096d112fdc639024c669.